### PR TITLE
Add Clang-Format configuration

### LINKED ISSUE #1441: Clang-format file
It would be helpful for developers if Z3's source tree came with a `.clang-format` file so that changes can be automatically formatted.

Note that clang-format can format whole files (usually not desirable) or it can format patches (i.e. with `git clang-format`) or regions of code (e.g. there's a vim plug-in that lets you format a selected region of text).

A handy tool for visually creating a `.clang-format` file can be found at https://zed0.co.uk/clang-format-configurator/ and documentation on various options can be found at https://clang.llvm.org/docs/ClangFormatStyleOptions.html

Could a `.clang-format` file be added that matches Z3's C++ style?

### PR DESCRIPTION
**Pull request by kbobyrev**
This patch introduces Clang-Format configuration and reformats a single source file to see the outcome of reformatting the sources.

Fixes #1441.

### CODE REVIEW

**Comment by kbobyrev** [comment]
There were few inconsistencies that I noticed in the code and I'm not really sure how to break a tie in some cases:

* `*` and `&` alignment: so far I've seen `int* ptr;` and `int * ptr;` with the first one being more common - this corresponds to `PointerAlignment: Left` Clang-Format option, but feel free to correct me if this assumption is wrong.
* Single-line if statements: I've seen both `if (condition) statement;` and `if (condition) \n statement;` (both are within several lines of code in the formatted file). I didn't investigate more than a couple of files but I think `AllowShortIfStatementsOnASingleLine: false` (disallowing one-liners) would be correct here.
* I didn't set`AlignConsecutiveDeclarations: true` which caused reformatting of class/field members but if I set this option then it is applied everywhere in the code which doesn't look nice from my point of view. Again, feel free to correect me if I'm wrong.

**Comment by kbobyrev** [comment]
The issue was fixed by the last commit, but the build is failing seemingly because of unrelated changes (sibling PRs fail with the same errors in the same configurations and pass in the same configurations).

File: src/solver/mus.cpp [hunk 1]
```diff
@@ -18,30 +18,29 @@ Module Name:
...
-    obj_map<expr, unsigned>  m_expr2lit;
-    model_ref                m_model;
-    expr_ref_vector          m_soft;
-    vector<rational>         m_weights;
-    rational                 m_weight;
-
-    imp(solver& s): 
-        m_solver(s), m(s.get_manager()), m_lit2expr(m),  m_assumptions(m), m_soft(m)
-    {}
+    solver& m_solver;
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: I would prefer these declarations to be mostly aligned as in the original file.

File: src/solver/mus.cpp [hunk 2]
```diff
@@ -50,16 +49,18 @@ struct mus::imp {
...
-        return is_uninterp_const(lit) || (m.is_not(lit, l) && is_uninterp_const(l));
+        expr* l;
+        return is_uninterp_const(lit) ||
+               (m.is_not(lit, l) && is_uninterp_const(l));
     }
-    
+
     unsigned add_soft(expr* lit) {
-        //SASSERT(is_literal(lit));
+        // SASSERT(is_literal(lit));
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: good

File: src/solver/mus.cpp [hunk 3]
```diff
@@ -50,16 +49,18 @@ struct mus::imp {
     }
 
     bool is_literal(expr* lit) const {
-        expr* l;		
-        return is_uninterp_const(lit) || (m.is_not(lit, l) && is_uninterp_const(l));
+        expr* l;
+        return is_uninterp_const(lit) ||
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: ok

File: src/solver/mus.cpp [hunk 4]
```diff
@@ -81,9 +82,12 @@ struct mus::imp {
         ptr_vector<expr> unknown(m_lit2expr.size(), m_lit2expr.c_ptr());
         expr_ref_vector core_exprs(m);
         TRACE("mus", m_solver.display(tout););
-        while (!unknown.empty()) { 
-            IF_VERBOSE(12, verbose_stream() << "(mus reducing core: " << unknown.size() << " new core: " << mus.size() << ")\n";);
-            TRACE("mus", display_vec(tout << "core:  ", unknown); display_vec(tout << "mus:   ", mus););
+        while (!unknown.empty()) {
+            IF_VERBOSE(12, verbose_stream()
+                               << "(mus reducing core: " << unknown.size()
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: not sure which editors support this alignment.

File: src/solver/mus.cpp [hunk 5]
```diff
@@ -113,10 +117,8 @@ struct mus::imp {
                         }
                     }
                     TRACE("mus", tout << "core exprs:" << core_exprs << "\n";
-                        display_vec(tout << "core:", unknown);
-                        display_vec(tout << "mus:", mus);
-                    );
-
+                          display_vec(tout << "core:", unknown);
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: OK

File: src/solver/mus.cpp [hunk 6]
```diff
@@ -142,33 +146,38 @@ struct mus::imp {
...
         bool min_core_valid = false;
         expr* min_lit = nullptr;
         while (!unknown.empty()) {
             expr* lit = unknown.back();
             unknown.pop_back();
             model_ref mdl;
-            scoped_append assume_mss(*this,  mus, mss);      // current satisfied literals
-            scoped_append assume_nmcs(*this, mus, nmcs);     // current non-satisfied literals
-            scoped_append assume_lit(*this,  mus, lit);      // current unknown literal
+            scoped_append assume_mss(*this, mus,
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: why does this break the line?

File: src/solver/mus.cpp [hunk 7]
```diff
@@ -142,33 +146,38 @@ struct mus::imp {
...
             switch (m_solver.check_sat(mus)) {
             case l_true: {
-                TRACE("mus", tout << "literal can be satisfied: " << mk_pp(lit, m) << "\n";);
+                TRACE("mus", tout << "literal can be satisfied: "
+                                  << mk_pp(lit, m) << "\n";);
                 mss.push_back(lit);
                 m_solver.get_model(mdl);
                 model_evaluator eval(*mdl.get());
-                for (unsigned i = 0; i < unknown.size(); ) {
+                for (unsigned i = 0; i < unknown.size();) {
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: ok

File: src/solver/mus.cpp [hunk 8]
```diff
@@ -220,13 +231,13 @@ struct mus::imp {
         for (expr* e : mss) {
             mss_set.insert(e);
         }
-        for (expr * e : min_core) {
+        for (expr* e : min_core) {
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: OK, fine to have * bind to the type. I don't like expr *e, which is very common in C/C++ code bases.
The reason is that I read * as part of the type expression. 
So expr and expr* are types, and 'e' is an expression of type expr*.

For that matter I also have developed a possibly peculiar convention of placing const qualifiers in post-fix mode.
I write T const & onstead of const T& so I don't have to recall what the const qualifier binds to.

File: src/solver/mus.cpp [hunk 9]
```diff
@@ -245,56 +256,56 @@ struct mus::imp {
         if (A.size() < B.size()) {
             expr_set::iterator it = A.begin(), end = A.end();
             for (; it != end; ++it) {
-                if (B.contains(*it)) return true;
+                if (B.contains(*it))
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: OK, fine to have this consistent

File: src/solver/mus.cpp [hunk 10]
```diff
@@ -245,56 +256,56 @@ struct mus::imp {
...
+                    return true;
             }
         }
         else {
             expr_set::iterator it = B.begin(), end = B.end();
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: Ups, I need to rewrite this to for (expr * e : B) ...

File: src/solver/mus.cpp [hunk 11]
```diff
@@ -245,56 +256,56 @@ struct mus::imp {
...
             fmls1.push_back(fml);
         }
         ~scoped_append() {
             m_fmls.shrink(m_size);
         }
     };
 
-    template<class T>
-    void display_vec(std::ostream& out, T const& v) const {
+    template <class T> void display_vec(std::ostream& out, T const& v) const {
```
Comment (NikolajBjorner) [inline src/solver/mus.cpp]: This seems arbitrary. Putting the template portion on a separate line shortens the next line and separates the template portion (polymorphism quantifier) from type signature.

File: src/solver/mus.h [hunk 12]
```diff
@@ -19,49 +19,52 @@ Module Name:
...
+       Assume that the solver context enforces that
        cls is equivalent to a disjunction of args.
-       Assume also that cls is a literal.           
+       Assume also that cls is a literal.
     */
     unsigned add_soft(expr* cls);
 
     void add_soft(unsigned sz, expr* const* clss) {
-        for (unsigned i = 0; i < sz; ++i) add_soft(clss[i]);
+        for (unsigned i = 0; i < sz; ++i)
```
Comment (NikolajBjorner) [inline src/solver/mus.h]: OK

**Comment by NikolajBjorner** [comment]
items 1, 2 seem right. For 3, in the pull request there is a realignment of field declarations. This reformatting makes reading what are the field names not so easy. I would prefer (insist) that it be possible to align them in a way that fits a tabular form whenever possible.

**Comment by NikolajBjorner** [comment]
closing as no reaction to fixing the realignment of field declarations. Fields are more readable when they are aligned. The other changes are fine.

**Pull request closed without merging**